File: api/src/db/schema.ts
import { pgTable, uuid, timestamp, text, integer, smallint, boolean, jsonb, real, varchar, bigint, index } from 'drizzle-orm/pg-core';

export const sessions = pgTable('sessions', {
  id: uuid('id').defaultRandom().primaryKey(),
  startedAt: timestamp('started_at', { withTimezone: true }).notNull(),
  endedAt: timestamp('ended_at', { withTimezone: true }),
  urlSeed: text('url_seed'),
  userAgent: text('user_agent'),
  metadata: jsonb('metadata'),
});

export const pages = pgTable('pages', {
  id: uuid('id').defaultRandom().primaryKey(),
  sessionId: uuid('session_id').references(() => sessions.id),
  url: text('url').notNull(),
  title: text('title'),
  navigatedAt: timestamp('navigated_at', { withTimezone: true }).notNull(),
  loadTimeMs: integer('load_time_ms'),
  statusCode: smallint('status_code'),
});

export const requests = pgTable('requests', {
  id: uuid('id').primaryKey(),
  pageId: uuid('page_id').references(() => pages.id),
  ts: timestamp('ts', { withTimezone: true }).notNull(),
  method: varchar('method', { length: 10 }),
  url: text('url').notNull(),
  urlHost: text('url_host'),
  resourceType: varchar('resource_type', { length: 32 }),
  initiatorType: varchar('initiator_type', { length: 32 }),
  headers: jsonb('headers'),
  postDataRef: text('post_data_ref'),
}, (table) => ({
  urlHostIdx: index('idx_requests_url_host').on(table.urlHost, table.ts),
  pageIdx: index('idx_requests_page').on(table.pageId, table.ts),
}));

export const responses = pgTable('responses', {
  requestId: uuid('request_id').references(() => requests.id),
  ts: timestamp('ts', { withTimezone: true }).notNull(),
  status: smallint('status'),
  statusText: varchar('status_text', { length: 64 }),
  headers: jsonb('headers'),
  bodyRef: text('body_ref'),
  bodySize: integer('body_size'),
  transferSize: integer('transfer_size'),
  mimeType: varchar('mime_type', { length: 128 }),
  timing: jsonb('timing'),
}, (table) => ({
  mimeIdx: index('idx_responses_mime').on(table.mimeType, table.ts),
}));

export const domEvents = pgTable('dom_events', {
  id: uuid('id').primaryKey(),
  pageId: uuid('page_id').references(() => pages.id),
  ts: timestamp('ts', { withTimezone: true }).notNull(),
  tsPageMs: real('ts_page_ms'),
  mutationType: varchar('mutation_type', { length: 32 }),
  targetPath: text('target_path'),
  payload: jsonb('payload'),
});

export const wsEvents = pgTable('ws_events', {
  id: uuid('id').primaryKey(),
  requestId: uuid('request_id').references(() => requests.id),
  ts: timestamp('ts', { withTimezone: true }).notNull(),
  direction: varchar('direction', { length: 8 }),
  opcode: smallint('opcode'),
  payloadRef: text('payload_ref'),
  masked: boolean('masked'),
});

export const screenshots = pgTable('screenshots', {
  id: uuid('id').primaryKey(),
  pageId: uuid('page_id').references(() => pages.id),
  ts: timestamp('ts', { withTimezone: true }).notNull(),
  trigger: varchar('trigger', { length: 32 }),
  format: varchar('format', { length: 8 }),
  width: smallint('width'),
  height: smallint('height'),
  fileSize: integer('file_size'),
  objectKey: text('object_key').notNull(),
  perceptualHash: bigint('perceptual_hash', { mode: 'number' }),
});

File: api-rust/src/db/schema.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Session {
    pub id: Uuid,
    pub started_at: DateTime<Utc>,
    pub ended_at: Option<DateTime<Utc>>,
    pub url_seed: Option<String>,
    pub user_agent: Option<String>,
    pub metadata: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Page {
    pub id: Uuid,
    pub session_id: Uuid,
    pub url: String,
    pub title: Option<String>,
    pub navigated_at: DateTime<Utc>,
    pub load_time_ms: Option<i32>,
    pub status_code: Option<i16>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Request {
    pub id: Uuid,
    pub page_id: Uuid,
    pub ts: DateTime<Utc>,
    pub method: Option<String>,
    pub url: String,
    pub url_host: Option<String>,
    pub resource_type: Option<String>,
    pub initiator_type: Option<String>,
    pub headers: Option<serde_json::Value>,
    pub post_data_ref: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Response {
    pub request_id: Uuid,
    pub ts: DateTime<Utc>,
    pub status: Option<i16>,
    pub status_text: Option<String>,
    pub headers: Option<serde_json::Value>,
    pub body_ref: Option<String>,
    pub body_size: Option<i32>,
    pub transfer_size: Option<i32>,
    pub mime_type: Option<String>,
    pub timing: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct DomEvent {
    pub id: Uuid,
    pub page_id: Uuid,
    pub ts: DateTime<Utc>,
    pub ts_page_ms: Option<f32>,
    pub mutation_type: Option<String>,
    pub target_path: Option<String>,
    pub payload: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct WsEvent {
    pub id: Uuid,
    pub request_id: Uuid,
    pub ts: DateTime<Utc>,
    pub direction: Option<String>,
    pub opcode: Option<i16>,
    pub payload_ref: Option<String>,
    pub masked: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Screenshot {
    pub id: Uuid,
    pub page_id: Uuid,
    pub ts: DateTime<Utc>,
    pub trigger: Option<String>,
    pub format: Option<String>,
    pub width: Option<i16>,
    pub height: Option<i16>,
    pub file_size: Option<i32>,
    pub object_key: String,
    pub perceptual_hash: Option<i64>,
}
